### PR TITLE
Add `TC006` rule for string literal types in `typing.cast()`

### LINKED ISSUE #114: Recommend quoting of type in `typing.cast()`
See https://twitter.com/zzzeek/status/1483093707637366795

For particularly complex casts there can be a significant performance hit, particularly in hot code paths, e.g.

```python
In [1]: %timeit list[tuple[bool | float | int | str, ...]] | None
451 ns ± 18.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [2]: %timeit "list[tuple[bool | float | int | str, ...]] | None"
4.01 ns ± 0.0235 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```

Given that the type in the cast is purely for type checkers there is probably no good reason why it shouldn't always be quoted.

### PR DESCRIPTION
**Pull request by ngnpope**
Closes #114.

### CODE REVIEW

**Comment by ngnpope** [comment]
Tried switching to `ast-compat` for backport of `ast.unparse()` required for Python < 3.9 instead of `astunparse`, but they both wrap expressions with extraneous parentheses. Will have to think what to do about this... 🤔

**Comment by ngnpope** [comment]
I found a replacement for both `astunparse` and `ast-compat` in `astor` which seems to produce similar output to `ast.unparse()`.

While it's not ideal having a dependency for this, with Python 3.11 nearly out it'll only be affecting Python 3.8, one of four supported versions. The only other alternative is to vendor [`Tools/parser/unparse.py`](https://github.com/python/cpython/blob/3.8/Tools/parser/unparse.py) from the Python 3.8 source tree, but I haven't checked whether it produces the expected output.

File: flake8_type_checking/checker.py [hunk 1]
```diff
@@ -409,6 +428,11 @@ def __init__(
         #: Set to the alias of typing if one is found
         self.typing_alias: Optional[str] = None
 
+        #: Where typing.cast() is called with an unquored type.
```
Comment (sondrelg) [inline flake8_type_checking/checker.py:431]:
```suggestion
        #: Where typing.cast() is called with an unquoted type.
```

**Review by sondrelg — COMMENTED** [review]
Looks good! 

I suppose we could drop the added dependency if we just change the error messages from specifying the annotation value to just telling the user what is wrong without specifying, but I'm not necessarily opposed to adding `astor` for 3.8. Just a thought, but do you think it would be weird to simplify the error message for 3.8 and have it output the annotation value where there's built-in support for it?

I'll see if I can get a chance to test the PR on a larger code base today and look over it one more time, then I'm happy to accept this if we don't find anything needing a tweak 🙂

**Comment by ngnpope** [comment]
> Just a thought, but do you think it would be weird to simplify the error message for 3.8 and have it output the annotation value where there's built-in support for it?

I guess we could simplify the error message for 3.8 only if the additional dependency is a problem. At least `astor` doesn't have any sub-dependencies so at least it's not pulling in a large tree of stuff. I did try the version from the CPython 3.8 branch, but it has the same issues with parentheses everywhere that `astunparse` does, so it's likely what was used to create that package in the first place. Anyway, that's a little more info, but I'll defer to what you want to do. 🙂 

> I'll see if I can get a chance to test the PR on a larger code base today and look over it one more time, then I'm happy to accept this if we don't find anything needing a tweak :slightly_smiling_face:

I gave it a quick go over a large code base and it looked sensible at a quick glance, but would be good to have a second opinion!

**Comment by sondrelg** [comment]
Tested and looks good! Thanks @ngnpope 👏

**Pull request merged**